src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicBuffers.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Microsoft.Quic;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal;
+
+/// <summary>
+/// Helper class to convert managed data into QUIC_BUFFER* consumable by MsQuic.
+/// It also allows reuse with repeated Reset/Initialize calls.
+/// Note that since this is struct and there's no finalizer, Dispose must be always called to release the unmanaged memory allocated by this struct.
+/// </summary>
+internal unsafe struct MsQuicBuffers : IDisposable
+{
+    // Handles to pinned memory blocks from the user.
+    private MemoryHandle[] _handles;
+    // Native memory block which holds the pinned memory pointers from _handles and can be passed to MsQuic as QUIC_BUFFER*.
+    private IntPtr _buffers;
+    // Number of QUIC_BUFFER instance currently allocated in _buffers, so that we can reuse the memory instead of reallocating.
+    private int _count;
+
+    public MsQuicBuffers()
+    {
+        _handles = Array.Empty<MemoryHandle>();
+        _buffers = IntPtr.Zero;
+        _count = 0;
+    }
+
+    public QUIC_BUFFER* Buffers => (QUIC_BUFFER*)_buffers;
+    public int Count => _count;
+
+    /// <summary>
+    /// The mothod initializes QUIC_BUFFER* with data from inputs, converted via toBuffer.
+    /// Note that the struct either needs to be freshly created or previously cleaned up with Reset.
+    /// </summary>
+    /// <param name="inputs">Inputs to get their byte array, pin them and pepare them to be passed to MsQuic as QUIC_BUFFER*.</param>
+    /// <param name="toBuffer">Method extracting byte array from the inputs, e.g. applicationProtocol.Protocol.</param>
+    /// <typeparam name="T">The type of the inputs.</typeparam>
+    public void Initialize<T>(IList<T> inputs, Func<T, ReadOnlyMemory<byte>> toBuffer)
+    {
+        if (_handles.Length < inputs.Count)
+        {
+            ArrayPool<MemoryHandle>.Shared.Return(_handles);
+            _handles = ArrayPool<MemoryHandle>.Shared.Rent(inputs.Count);
```

Comment:
`_handles` array that came from ArrayPool.Rent is not guaranteed to be zero-initialized. For example, it may contain handles from previous use that were disposed already.

If there is an exception (e.g. from the alloc call below) before the array is initialized, you may end up disposing these random handles later and very hard to debug crash.

I think the easiest way to deal with this is by zero initializing the rented array.

Reply:
If the loop in `Reset` gets fixed then we shouldn't need zeroing, since `_count` is set after native alloc. I'm even thinking of moving it inside the loop to correspond to currently pinned items in all cases.